File: skills_manager/resolver.py
"""Manifest resolution."""
from __future__ import annotations

from pathlib import Path
from typing import Any

from . import store

SCOPES = ("global", "profile", "project", "session")


def _manifest_sequence(project: str | Path | None = None) -> list[tuple[str, dict[str, Any]]]:
    seq = []
    for scope in SCOPES:
        if scope == "project" and project is None:
            continue
        seq.append((scope, store.load_manifest(scope, project=project)))
    return seq


def _apply_list(effective: set[str], values: list[str], op: str, reasons: dict[str, list[str]], reason: str) -> None:
    for value in values:
        if op == "enable":
            effective.add(value)
        else:
            effective.discard(value)
        reasons.setdefault(value, []).append(reason)


def resolve(client: str, project: str | Path | None = None) -> dict[str, Any]:
    managed = store.all_skills()
    effective: set[str] = set()
    reasons: dict[str, list[str]] = {}
    for scope, manifest in _manifest_sequence(project):
        _apply_list(effective, list(manifest.get("enable", [])), "enable", reasons, f"{scope}:enable")
        _apply_list(
            effective,
            list(manifest.get("clients", {}).get(client, {}).get("enable", [])),
            "enable",
            reasons,
            f"{scope}:{client}:enable",
        )
        _apply_list(effective, list(manifest.get("disable", [])), "disable", reasons, f"{scope}:disable")
        _apply_list(
            effective,
            list(manifest.get("clients", {}).get(client, {}).get("disable", [])),
            "disable",
            reasons,
            f"{scope}:{client}:disable",
        )
    desired: dict[str, dict[str, Any]] = {}
    unknown = sorted(skill_id for skill_id in effective if skill_id not in managed)
    for skill_id in sorted(effective):
        meta = managed.get(skill_id)
        if not meta:
            continue
        if not meta.get("compatibility", {}).get(client, False):
            reasons.setdefault(skill_id, []).append(f"{client}:incompatible")
            continue
        alias = meta.get("aliases", {}).get(client) or skill_id.split(".")[-1]
        desired[skill_id] = {"id": skill_id, "alias": alias, "meta": meta, "reasons": reasons.get(skill_id, [])}
    by_skill = {}
    for skill_id, meta in managed.items():
        by_skill[skill_id] = {
            "enabled": skill_id in desired,
            "alias": meta.get("aliases", {}).get(client),
            "compatible": meta.get("compatibility", {}).get(client, False),
            "reasons": reasons.get(skill_id, []),
        }
    return {"client": client, "desired": desired, "skills": by_skill, "unknown_enabled_ids": unknown}


def set_skill(scope: str, skill_id: str, enabled: bool, client: str = "all", project: str | Path | None = None) -> Path:
    manifest = store.load_manifest(scope, project=project)
    if client == "all":
        target = manifest
    else:
        target = manifest.setdefault("clients", {}).setdefault(client, {"enable": [], "disable": []})
    add_key, remove_key = ("enable", "disable") if enabled else ("disable", "enable")
    target.setdefault(add_key, [])
    target.setdefault(remove_key, [])
    if skill_id not in target[add_key]:
        target[add_key].append(skill_id)
    target[remove_key] = [x for x in target[remove_key] if x != skill_id]
    return store.save_manifest(scope, manifest, project=project)

File: src/core/resolver.ts
import { resolve as resolvePath } from "node:path";
import { appendAction } from "./action-log.js";
import { allSkills, loadManifest, saveManifest, type SkillMeta } from "./store.js";
import type { Client } from "./schemas.js";
import type { Env } from "./paths.js";

export const SCOPES = ["global", "project", "session"] as const;
export type ResolverScope = (typeof SCOPES)[number];
export type DesiredSkill = {
  id: string;
  alias: string;
  meta: SkillMeta;
  reasons: string[];
};

export type ResolvedSkillSummary = {
  enabled: boolean;
  alias?: string;
  compatible: boolean;
  reasons: string[];
};

export async function resolveSkillRef(ref: string, env: Env = process.env): Promise<string> {
  if (ref.startsWith("skill.")) {
    return ref;
  }
  const managed = await allSkills(env);
  const matches = Object.entries(managed)
    .filter(([skillId, meta]) => ref === skillId.split(".").at(-1) || Object.values(meta.aliases).includes(ref))
    .map(([skillId]) => skillId);
  return matches.length === 1 ? matches[0]! : ref;
}

function applyList(effective: Set<string>, values: string[], op: "enable" | "disable", reasons: Map<string, string[]>, reason: string): void {
  for (const value of values) {
    if (op === "enable") {
      effective.add(value);
    } else {
      effective.delete(value);
    }
    const existing = reasons.get(value) ?? [];
    existing.push(reason);
    reasons.set(value, existing);
  }
}

async function manifestSequence(options: { project?: string; env?: Env }): Promise<Array<[ResolverScope, Awaited<ReturnType<typeof loadManifest>>]>> {
  const sequence: Array<[ResolverScope, Awaited<ReturnType<typeof loadManifest>>]> = [];
  for (const scope of SCOPES) {
    if (scope === "project" && !options.project) {
      continue;
    }
    sequence.push([scope, await loadManifest(scope, manifestOptions(options))]);
  }
  return sequence;
}

function manifestOptions(options: { project?: string; env?: Env }): { project?: string; env?: Env } {
  return {
    ...(options.project ? { project: options.project } : {}),
    ...(options.env ? { env: options.env } : {})
  };
}

export async function resolveDesired(client: Client, options: { project?: string; env?: Env } = {}): Promise<{
  client: Client;
  desired: Record<string, DesiredSkill>;
  skills: Record<string, ResolvedSkillSummary>;
  unknown_enabled_ids: string[];
}> {
  const managed = await allSkills(options.env);
  const effective = new Set<string>();
  const reasons = new Map<string, string[]>();

  for (const [scope, manifest] of await manifestSequence(options)) {
    applyList(effective, manifest.enable, "enable", reasons, `${scope}:enable`);
    applyList(effective, manifest.clients[client].enable, "enable", reasons, `${scope}:${client}:enable`);
    applyList(effective, manifest.disable, "disable", reasons, `${scope}:disable`);
    applyList(effective, manifest.clients[client].disable, "disable", reasons, `${scope}:${client}:disable`);
  }

  const desired: Record<string, DesiredSkill> = {};
  const unknown_enabled_ids = [...effective].filter((skillId) => !managed[skillId]).sort();
  for (const skillId of [...effective].sort()) {
    const meta = managed[skillId];
    if (!meta) {
      continue;
    }
    if (!meta.compatibility[client]) {
      reasons.set(skillId, [...(reasons.get(skillId) ?? []), `${client}:incompatible`]);
      continue;
    }
    const alias = meta.aliases[client] || skillId.split(".").at(-1) || skillId;
    desired[skillId] = { id: skillId, alias, meta, reasons: reasons.get(skillId) ?? [] };
  }

  const skills: Record<string, ResolvedSkillSummary> = {};
  for (const [skillId, meta] of Object.entries(managed)) {
    const alias = meta.aliases[client];
    skills[skillId] = {
      enabled: skillId in desired,
      ...(alias === undefined ? {} : { alias }),
      compatible: meta.compatibility[client],
      reasons: reasons.get(skillId) ?? []
    };
  }

  return { client, desired, skills, unknown_enabled_ids };
}

export async function setSkill(
  scope: ResolverScope,
  skillId: string,
  enabled: boolean,
  options: { client?: Client | "all"; project?: string; surface?: string; env?: Env } = {}
): Promise<string> {
  const client = options.client ?? "all";
  const manifest = await loadManifest(scope, manifestOptions(options));
  const target = client === "all" ? manifest : manifest.clients[client];
  const addKey = enabled ? "enable" : "disable";
  const removeKey = enabled ? "disable" : "enable";

  if (!target[addKey].includes(skillId)) {
    target[addKey].push(skillId);
  }
  target[removeKey] = target[removeKey].filter((value) => value !== skillId);

  const path = await saveManifest(scope, manifest, manifestOptions(options));
  await appendAction(
    enabled ? "enable" : "disable",
    {
      surface: options.surface ?? "core",
      scope,
      client,
      skill_id: skillId,
      manifest_path: path,
      project_path: options.project ? resolvePath(options.project) : null
    },
    options.env
  );
  return path;
}
